libcloudforensics/providers/kubernetes/base.py

```diff
@@ -0,0 +1,226 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Kubernetes core class structure."""
+
+import abc
+from typing import List, TypeVar, Callable, Optional
+
+from kubernetes import client
+
+from libcloudforensics.providers.kubernetes import selector
+
+
+class K8sClient(metaclass=abc.ABCMeta):
+  """Abstract class representing objects that use the Kubernetes API."""
+
+  T = TypeVar('T')
+
+  def __init__(self, api_client: client.ApiClient) -> None:
+    """Creates an object holding Kubernetes API client.
+
+    Args:
+      api_client (client.ApiClient): The authenticated Kubernetes API client to
+        the cluster.
+    """
+    self._api_client = api_client
+
+  def _Api(self, api_class: Callable[[client.ApiClient], T]) -> T:
+    """Given an API class, creates an instance with the authenticated client.
+
+    Example usage:
+    ```
+    nodes = self._Api(CoreV1Api).list_node()
+    ```
+
+    Args:
+      api_class (Callable[[ApiClient], T]): The API class.
+
+    Returns:
+      T: An authenticated instance of the desired API class.
+    """
+    return api_class(self._api_client)
+
+
+class K8sResource(K8sClient, metaclass=abc.ABCMeta):
+  """Abstract class representing a Kubernetes resource.
+
+  Attributes:
+    name (str): The name of this resource.
+  """
+
+  def __init__(self, api_client: client.ApiClient, name: str) -> None:
+    """Creates a Kubernetes resource holding Kubernetes API client.
+
+    Args:
+      api_client (ApiClient): The authenticated Kubernetes API client to
+        the cluster.
+      name (str): The name of this resource.
+    """
+    super().__init__(api_client)
+    self.name = name
+
+  @abc.abstractmethod
+  def Read(self) -> object:
+    """Returns the resulting read operation for the resource.
+
+    For example, a Node resource would call CoreV1Api.read_node, a Pod
+    resource would call CoreV1Api.read_namespaced_pod. The return values
+    of these read calls do not share a base class, hence the object return
+    type.
+
+    Returns:
+      object: The result of this resource's matching read operation.
+    """
+
+
+class K8sCluster(K8sClient):
```

Comment:
maybe add __init__(self, api_client) function here that calls super().__init__(api_client) and make check that api_client is authenticated. This will also make it clearer

Reply:
This causes a pylint warning, useless-super-delegation

Reply:
ok, seems fine to remove from here, you can either add a check that apiClient is authenticated in k8sClient or discard it completely in case an authenticated client isn't needed for some API calls